src/Processors/QueryPlan/Optimizations/useVectorSearch.cpp

```diff
@@ -371,10 +375,18 @@ bool optimizeVectorSearchSecondPass(QueryPlan::Node & /*root*/, Stack & stack, Q
             /// Remove the physical vector column from ReadFromMergeTreeStep, add virtual "_distance" column
             read_from_mergetree_step->replaceVectorColumnWithDistanceColumn(search_column);
 
+            const bool need_cast = !WhichDataType(sort_column_node->result_type).isFloat32();
+
             /// Now replace the "cosineDistance(vec, [1.0, 2.0...])" node in the DAG by the "_distance" node
             expression.removeUnusedResult(sort_column); /// Removes the OUTPUT cosineDistance(...) FUNCTION Node
             expression.removeUnusedActions(); /// Removes the vector column INPUT node (it is no longer needed)
-            const auto * distance_node = &expression.addInput("_distance",std::make_shared<DataTypeFloat64>());
+            const auto * distance_node = &expression.addInput("_distance",std::make_shared<DataTypeFloat32>());
+
+            /// The virtual column is Float32 type, but the cosineDistance/L2Distance could have return type
+            /// of Float64 or Float32, hence add a cast. The MergeTreeRangeReader can only return Float32 column.
+            if (need_cast)
+                distance_node = &expression.addCast(*distance_node, sort_column_node->result_type, "_CAST_distance");
```

Comment:
Not sure but will it make sense to cast up to Float64 instead? The precision might be a tiny bit better at a negligible higher memory consumption. But I don't have a strong opionion here.